Migrate to doctrine/lexer v3.0

Ph0tonic opened the pull request:
[Doctrine/L](https://github.com/doctrine/lexer) just released a brand new version 3.0 and here is a PR to use this newer version.

Let me know @egulias if I need to change anything, I am not familiar with psalm annotations.

Ph0tonic committed:
Migrate to doctrine/lexer v2.0

egulias commented on src/EmailLexer.php:
```diff
@@ -131,16 +132,16 @@ class EmailLexer extends AbstractLexer
-     * @var array
+     * @var Token
      *
      * @psalm-var array{value:string, type:null|int, position:int}|array<empty, empty>
```
You can remove this annotation, given `$previous` is now a proper type

egulias commented on src/EmailLexer.php:
```diff
@@ -131,16 +132,16 @@ class EmailLexer extends AbstractLexer
+     * @var object
      *
      * @psalm-suppress NonInvariantDocblockPropertyType
      * @psalm-var array{value:string, type:null|int, position:int}
```
You can remove this annotation, given `$previous` is now a proper type

egulias commented on src/EmailLexer.php:
```diff
@@ -131,16 +132,16 @@ class EmailLexer extends AbstractLexer
      * The last matched/seen token.
      *
-     * @var array
+     * @var object
```
Should this also be `Token`?

egulias commented on src/EmailLexer.php:
```diff
@@ -156,11 +157,7 @@ class EmailLexer extends AbstractLexer
     public $lookahead;
 
     /** @psalm-var array{value:'', type:null, position:0} */
```
You can remove this annotation, given `$previous` is now a proper type.

egulias commented on src/EmailLexer.php:
```diff
@@ -257,37 +255,37 @@ protected function getType(&$value)
      * getPrevious
      *
-     * @return array
+     * @return object
```
Replace with `Token`

egulias commented on src/EmailLexer.php:
```diff
@@ -257,37 +255,37 @@ protected function getType(&$value)
+     * @return object
      */
-    public function getPrevious() : array
+    public function getPrevious(): object
```
Type is `Token` here, right?

egulias commented on src/MessageIDParser.php:
```diff
@@ -25,19 +25,19 @@ class MessageIDParser extends Parser
     {
         if (!$this->hasAtToken()) {
-            return new InvalidEmail(new NoLocalPart(), $this->lexer->token["value"]);
+            return new InvalidEmail(new NoLocalPart(), $this->lexer->value);
```
`$this->lexer->token->value`

egulias reviewed (changes requested):
Hey, thanks for taking the time. Very nice PR to update the dependency, at last it brings a `Token` type.
I have made a bunch of comments but please take a close look to Scrutinizer comments on the files. There are methods in the new version of the Lexer library that have types different to the one in the code.
Also check the _Checks_ tab and see why tests are failing.

Looking forward to the changes. And thanks again for the PR!

Ph0tonic committed:
Apply feedbacks

Ph0tonic committed:
Fix Tests + Psalm issues

Ph0tonic commented:
Hey @egulias ,

Thank for your feedback, I applied them and fixed the tests. There was a type check issue due to the value `null` of the lex `S_EMPTY`. I changed it to `-1` and made the relevant changes to make it work.

I also finally managed to fix all Psalm issues so I'll let you review the changes 👍

Pasqualle commented:
doctrine/lexer v3.0.0 is also available now:
https://packagist.org/packages/doctrine/lexer

Ph0tonic committed:
Upgrade to doctrine/lexer 3.x

Ph0tonic commented:
> doctrine/lexer v3.0.0 is also available now: https://packagist.org/packages/doctrine/lexer

Thank for the info @Pasqualle 

@egulias The main difference with this new v3.0.0 version is the dropped support for PHP <8.1. From what I have read in your readme, we should create a new v4.x release branch.
I updated PHP versions in Travis, with the newer supported version but I wasn't sure about the support of ppc64le architecture. Let me know what you think about that. 👍

Ph0tonic committed:
Use php nullsafe operator  to fix type issues

Ph0tonic committed:
Fix types issues

Ph0tonic committed:
Fix psalm type issues info

Ph0tonic committed:
Fix psalm type check

derrabus commented:
Note that Lexer 3 is Lexer 2 minus the deprecation layer, so requiring Lexer 3 without allowing v2 as well is not really necessary. Lexer 3 is not supported by Doctrine ORM 2 and it likely won't ever be. This basically means that bumping to `"doctrine/lexer": "^3"` instead of `"doctrine/lexer": "^2 || ^3"` would forbid the use of this package in a project that also uses Doctrine ORM.

derrabus commented:
I've submitted #340 to unblock Lexer 2 (and doctrine/annotations 2) in Symfony's CI.

egulias commented:
Thanks @derrabus, very useful.
Will merge #340 and then upgrade this library to v4 with this code.
Will work on this during the week.

@Ph0tonic I might need you to do a PR to another branch for this. Will let you know.

Thanks guys!

Ph0tonic commented:
> Thanks @derrabus, very useful. Will merge #340 and then upgrade this library to v4 with this code. Will work on this during the week.
> 
> @Ph0tonic I might need you to do a PR to another branch for this. Will let you know.
> 
> Thanks guys!

Great thank you, let me know if you need me to do anything.

egulias commented on README.md:
```diff
@@ -21,16 +21,17 @@ This library aims to support RFCs:
+| ------- | :--------: | :-----: | :---------------------: | :---: |
+| v4.x    | 2022/12/20 |    -    |            X            |   X   |
+| v3.x    | 2020/12/29 |    -    |            X            |   X   |
+| v2.1.x  | 2016/05/16 | 01/2022 |            X            |       |
```
Can you please mark v2 as EOL -> YES
And keep release date for 4 clear for now, work in progress!

egulias commented on composer.json:
```diff
@@ -1,20 +1,28 @@
-  "homepage":     "https://github.com/egulias/EmailValidator",
-  "keywords":     ["email", "validation", "validator", "emailvalidation", "emailvalidator"],
-  "license":      "MIT",
+  "name": "egulias/email-validator",
```
Please keep the original formatting :)

egulias commented on composer.json:
```diff
@@ -1,20 +1,28 @@
-    "doctrine/lexer": "^1.2",
+    "php": ">=8.1",
+    "doctrine/lexer": "^3.0",
     "symfony/polyfill-intl-idn": "^1.15"
```
Please upgrade to latest, 1.26

egulias commented on .travis.yml:
```diff
@@ -2,16 +2,12 @@ language: php
+    - php: 8.1
       env:
         - psalm=yes
       dist: bionic
```
Update to Jammy please, this was set like this to be able to build PHP 7.

egulias reviewed (changes requested):
We are almost there! A couple of more changes please.
I'll create the 4.x branch and will request you to do the PR to that branch, once we have sorted out all the changes on this PR (so keep the conversation in one place).

Thanks!

Ph0tonic committed:
Fix readme + clean composer.json

Ph0tonic commented:
Great thank you for your feedback. Here are the requested changes. 👍

derrabus commented on composer.json:
```diff
@@ -1,21 +1,21 @@
-    "doctrine/lexer": "^1.2",
-    "symfony/polyfill-intl-idn": "^1.15"
+    "php": ">=8.1",
+    "doctrine/lexer": "^3.0",
```
Again, I would strongly recommend to loosen this constraint to…

```suggestion
    "doctrine/lexer": "^2.0 || ^3.0",
```

This should be possible without much extra effort and avoid dependency conflicts downstream for the time being.

egulias replied:
I have checked Doctrine's ORM and v3 is requiring v2 of the lexer, so @Ph0tonic can you please update to 
```suggestion
    "doctrine/lexer": "^2 || ^3",
```
Thanks @derrabus

Ph0tonic replied:
Done 👍

Ph0tonic committed:
loosen constraint over lexer

egulias commented:
@Ph0tonic please close this PR and create a new one against v4 branch: https://github.com/egulias/EmailValidator/tree/4.x .
I'll take care of the conflicts.
Thnkas!

egulias commented:
Closing in favor of #344

derrabus commented:
You mean #344, I guess. 🙃

egulias commented:
Updated, thanks!
